File: app/src/main/java/com/ryanrvldo/moviecatalogue/viewmodel/FavoritesViewModel.java
package com.ryanrvldo.moviecatalogue.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ryanrvldo.moviecatalogue.data.Repository;
import com.ryanrvldo.moviecatalogue.data.source.model.Movie;
import com.ryanrvldo.moviecatalogue.data.source.model.TvShow;

import java.util.List;

public class FavoritesViewModel extends ViewModel {

    private Repository repository;
    private LiveData<List<Movie>> favoriteMovies;
    private LiveData<List<TvShow>> favoriteTvShows;

    public FavoritesViewModel(Repository mRepository) {
        this.repository = mRepository;
    }

    public void deleteFavMovie(Movie movie) {
        repository.deleteFavMovie(movie);
        favoriteMovies = repository.getFavMovies();
    }

    public void setFavoriteMovies() {
        favoriteMovies = new MutableLiveData<>();
        favoriteMovies = repository.getFavMovies();
    }

    public LiveData<List<Movie>> getFavoriteMovies() {
        if (favoriteMovies == null) {
            setFavoriteMovies();
        }
        return favoriteMovies;
    }

    public void insertFavoriteTvShow(TvShow tvShow) {
        repository.insertFavTv(tvShow);
        favoriteTvShows = repository.getFavTvShows();
    }

    public TvShow getFavTv(int tvShowId) {
        return repository.getFavTv(tvShowId);
    }

    public void deleteFavTv(TvShow tvShow) {
        repository.deleteFavTv(tvShow);
        favoriteTvShows = repository.getFavTvShows();
    }

    public void setFavoriteTvShows() {
        favoriteTvShows = new MutableLiveData<>();
        favoriteTvShows = repository.getFavTvShows();
    }

    public LiveData<List<TvShow>> getFavoriteTvShows() {
        if (favoriteTvShows == null) {
            setFavoriteTvShows();
        }
        return favoriteTvShows;
    }
}

File: app/src/main/java/com/ryanrvldo/moviecatalogue/ui/viewmodel/FavoritesViewModel.kt
package com.ryanrvldo.moviecatalogue.ui.viewmodel

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ryanrvldo.moviecatalogue.data.Repository
import com.ryanrvldo.moviecatalogue.data.model.Movie
import com.ryanrvldo.moviecatalogue.data.model.TvShow
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class FavoritesViewModel @ViewModelInject constructor(
    private val repository: Repository
) : ViewModel() {

    val favoriteMovies: LiveData<List<Movie>> by lazy {
        repository.getFavoriteMovies()
    }

    val favoriteTvShows: LiveData<List<TvShow>> by lazy {
        repository.getFavoriteTvShows()
    }

    fun deleteFavMovie(movie: Movie) = viewModelScope.launch(Dispatchers.IO) {
        repository.removeFavoriteMovie(movie)
    }

    fun deleteFavTv(tvShow: TvShow) = viewModelScope.launch(Dispatchers.IO) {
        repository.removeFavoriteTvShow(tvShow)
    }
}